---
title: "Multilocus estimates of pi (infinite sites)"
output: html_notebook
---

# Overview

We'll simulate $L$ independently inherited SNPs, using an infinite sites model and examining only those samples that have only one segregating site. We'll estimate $\theta$ Ewen's sampling formula, $\theta_B$, where we assume the same $\theta_B$ across all loci. There's no point in trying $\theta_k$, since it will always give the same answer for a particular sample size.

## Load libraries and clear out memory

```{r}
library(phyclust)
library(rstan)
library(tidyverse)
library(ggplot2)

rm(list = ls())

options(mc.cores = parallel::detectCores())
rstan_options(auto_write = TRUE)
```

## Utility functions

```{r}
## construct matrix of allele counts, loci in rows
##
construct_a <- function(ms_tree, n_sample, seg_sites) {
  a <- numeric(n_sample)
  if (seg_sites > 0) {
    counts <- table(ms_tree[5:(n_sample+4)])
    for (count in 1:length(counts)) {
      index <- counts[count]
      a[index] <- a[index] + 1
    }
  } else {
    a[n_sample] <- 1
  }
  return(a)
}

## Run the simulation for an infinite sites model
##
simulate_infinite <- function(n_sample, n_loci, theta, 
                              n_rep = 1, shape = 1, rate = 1) 
{
  theta_B <- numeric(n_rep)
  for (rep in 1:n_rep) {
    if (n_rep > 1) {
      cat(".")
      if (rep %% 5 == 0) {
        count <- sprintf("%4d", rep)
        cat(count)
      }
      if (rep %% 50 == 0) {
        cat("\n")
      }
    }
    ## simulate the data
    ##
    a <- matrix(nrow = n_loci, ncol = n_sample)
    for (l in 1:n_loci) {
      seg_sites <- -1
      while (seg_sites != 1) {
        ms_tree <- phyclust::ms(nsam = n_sample, nreps = 1, opts = paste("-t", theta))
        seg_sites <- as.numeric(gsub(".* ([0-9]+)", "\\1", ms_tree[3]))
      }
      ## construct a matrix for Bayesian estimate
      ##
      a[l, ] <- construct_a(ms_tree, n_sample, seg_sites)
    }
    ## Bayesian estimate: single theta shared across all loci
    ##
    stan_data <- list(N = n_sample,
                      L = n_loci,
                      a = a,
                      shape = shape,
                      rate = rate) 
    fit <- stan(file = "ewens-multilocus.stan",
                data = stan_data,
                refresh = 0)
    fit_df <- as.data.frame(fit)
    theta_B[rep] <- mean(fit_df$theta, na.rm = TRUE)
    #cat(theta_k[rep], " ", theta_B[rep], "\n")
  }
  ret_val <- tibble(theta_B = theta_B,
                    eps_theta_B = theta - theta_B)
  return(ret_val)
}
```

## Run a simulation

```{r}
results <- simulate_infinite(n_sample = 25,
                             n_loci = 100,
                             theta = 0.1,
                             n_rep = 10)

cat("theta_B: ", mean(results$theta_B), "(", sqrt(mean(results$eps_theta_B^2)), ")", "\n", sep = "")

for_plot <- pivot_longer(results, cols = everything(), names_to = "Model") %>%
  filter(Model == "theta_k" | Model == "theta_B")
p <- ggplot(for_plot, aes(x = value, fill = Model, color = Model)) +
  geom_density(alpha = 0.2) +
  geom_vline(xintercept = 0.1, color = "salmon", linetype = "dashed") +
  theme_bw()
p
```

Using SNPs with this approach is clearly a bad idea, unless maybe you simulate short sequences and place mutations with seqgen. Let's see how many segregating sites in simulations.

```{r}
n_sample <- 25
seq_len <- 250
theta <- 0.0001

seg_sites <- numeric(100)
for (i in 1:100) {
      ms_tree <- phyclust::ms(nsam = n_sample, nreps = 1, opts = paste("-T"))
      ## N.B.: seqs_chr comes back in PHYLIP format as a vector of strings (including the taxon)
      ## HKY without -t (transitiion/transversion ratio) and without -f (frequencies) makes the
      ## model Jukes-Cantor
      ##
      options <- paste("-mHKY -l", seq_len, " -d", theta, " -q", sep = "")
      seqs_chr <- seqgen(opts = options, newick.tree = ms_tree[3])
      ## We write the result to disk
      ##
      write(seqs_chr, "seqs.phy")
      ## read it back using read.dna (from ape) and delete it
      ##
      tmp <- read.dna("seqs.phy")
      file.remove("seqs.phy")
      seg_sites[i] <- length(seg.sites(tmp))
}

for_plot <- tibble(seg_sites = seg_sites)
p <- ggplot(for_plot, aes(x = seg_sites)) +
  geom_histogram() +
  theme_bw()
p
```

You can get it down to a small number of segregating sites, but how much information will you have? This is just made up biallelic SNP data.

```{r}
    n_sample <- 25
    n_loci <- 100
    shape <- 1
    rate <- 1
    
    a <- matrix(data = rep(0, n_loci*n_sample), nrow = n_loci, ncol = n_sample)
    for (i in 1:n_loci) {
      n <- rbinom(1, 25, 0.2)
      a[i, n] <- 1
      a[i, 25 - n] <- 1
    }    
    
    stan_data <- list(N = n_sample,
                      L = n_loci,
                      a = a,
                      shape = shape,
                      rate = rate) 
    fit <- stan(file = "ewens-multilocus.stan",
                data = stan_data,
                refresh = 0)
    fit_df <- as.data.frame(fit)
    
    p <- ggplot(fit_df, aes(x = theta)) +
      geom_density() +
      theme_bw()
    p
```

There seems to be a reasonable amount of information there. Maybe this is worth pursuing. Now the question is whether there's a more efficient way to generate SNP samples.

## ESF conditional on polymorphism

I'm going to put off efficicent SNP generation for a while. I want to investigate how large the difference is between estimates of $\theta$ based on the full ESF and those conditional on polymorphism. SNPs are a conditional sample, and estimates made without taking that into account are likely to be biased.

```{r}
construct_a_finite <- function(dna_seqs, n_sample) {
  a <- numeric(n_sample)
  if (length(seg.sites(dna_seqs)) > 0 ) {
    n_seqs <- nrow(dna_seqs)
    dna_seqs_chr <- character(n_seqs)
    for (i in 1:n_seqs) {
      dna_seqs_chr[i] <- paste(dna_seqs[i, ], collapse = "")
    }
    ##
    ## need to figure out how to construct table of haplotype counts from table of 
    ## sequences
    ##
    counts <- table(as.factor(dna_seqs_chr))
    for (count in 1:n_sample) {
      index <- counts[count]
      a[index] <- a[index] + 1
    }
  } else {
    a[n_sample] <- 1
  }
  return(a)
}

## Run the simulation for a finite length sequence
##
## Note: ts_tv = 0.5 gives equal rates since there are 4 types of transversions and only 
## 2 types of transitions
##
simulate_finite <- function(n_sample, seq_len, n_loci, theta,  
                            n_rep = 1, shape = 1, rate = 1, conditional = FALSE,
                            ts_tv = 0.5) 
{
  theta_B <- numeric(n_rep)
  for (rep in 1:n_rep) {
    if (n_rep > 1) {
      cat(".")
      if (rep %% 5 == 0) {
        count <- sprintf("%4d", rep)
        cat(count)
      }
      if (rep %% 50 == 0) {
        cat("\n")
      }
    }
    ## simulate the data
    ##
    a <- matrix(nrow = n_loci, ncol = n_sample)
    for (l in 1:n_loci) {
      seg_sites <- 0
      ## if conditional == TRUE, keep constructing samples until we have a SNP
      ## Note: as written there could be more than two alleles at the SNP
      ##
      while ((seg_sites < 1)) {
        ms_tree <- phyclust::ms(nsam = n_sample, nreps = 1, opts = paste("-T"))
        ## N.B.: seqs_chr comes back in PHYLIP format as a vector of strings (including the taxon)
        ## HKY with -t0.5 and without -f (frequencies) makes the model Jukes-Cantor
        ##
        options <- paste("-mHKY -l", seq_len, " -d", theta, " -t", ts_tv, " -q", sep = "")
        seqs_chr <- seqgen(opts = options, newick.tree = ms_tree[3])
        ## We write the result to disk
        ##
        write(seqs_chr, "seqs.phy")
        ## read it back using read.dna (from ape) and delete it
        ##
        tmp <- read.dna("seqs.phy")
        file.remove("seqs.phy")
        ## estimate theta using Watterson approach (theta_k)
        ##
        seg_sites <- length(seg.sites(tmp))
        if (conditional == FALSE) {
          break;
        }
      }
      ## construct a matrix for Bayesian estimate
      ##
      a[l, ] <- construct_a_finite(tmp, n_sample)
    }
    ## Bayesian estimate: single theta shared across all loci
    ##
    stan_data <- list(N = n_sample,
                      L = n_loci,
                      a = a,
                      shape = shape,
                      rate = rate) 
    ## Note: In the first simulation where I defined conditional == TRUE, I commented
    ## out the lines that would call the conditional Stan code.
    ##
    if (conditional) {
      model <- "ewens-multilocus-conditional.stan"
    } else {
      model <- "ewens-multilocus.stan"
    }
    fit <- stan(file = model,
                data = stan_data,
                refresh = 0)
    fit_df <- as.data.frame(fit)
    theta_B[rep] <- mean(fit_df$theta, na.rm = TRUE)/seq_len
  }
  ret_val <- tibble(theta_B = theta_B)
  return(ret_val)
}
```

Suppressing warnings because there are a small number of divergent transitions after warmup when theta is very small.

```{r warning = FALSE}
results <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, n_rep = 100,
                           shape = 1, rate = 1, conditional = FALSE)

p <- ggplot(results, aes(x = theta_B)) +
  geom_density() +
  geom_vline(xintercept = 0.0001, color = "salmon", linetype = "dashed") +
  theme_bw() +
  theme(axis.text.y = element_blank())
p
```
That was allowing for cases without polymorphism. Now we'll try it restricting it to cases where there is polymorphism.

```{r warning = FALSE}
results <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, n_rep = 100,
                           shape = 1, rate = 1, conditional = TRUE)

p <- ggplot(results, aes(x = theta_B)) +
  geom_density() +
  geom_vline(xintercept = 0.0001, color = "salmon", linetype = "dashed") +
  theme_bw() +
  theme(axis.text.y = element_blank())
p
```

Clearly estimates based on SNPs that don't take into account ascertainment bias won't work well. This isn't surprising, but I'm glad that I checked. Note: the conditional stan model was commented out for that simulation. I've removed the comment for the simulation below.

```{r}
results <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, n_rep = 100,
                           shape = 1, rate = 1, conditional = TRUE)

p <- ggplot(results, aes(x = theta_B)) +
  geom_density() +
  geom_vline(xintercept = 0.0001, color = "salmon", linetype = "dashed") +
  theme_bw() +
  theme(axis.text.y = element_blank())
p
```

There appears to be more uncertainty in the estimate than when sampling unconditionally, but it appears that conditioning works. Let's see how simulating sequences with the HKY model and a transition/transversion ratio of 3:1 affects the results.


```{r}
results <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, n_rep = 100,
                           shape = 1, rate = 1, conditional = TRUE, ts_tv = 3)

p <- ggplot(results, aes(x = theta_B)) +
  geom_density() +
  geom_vline(xintercept = 0.0001, color = "salmon", linetype = "dashed") +
  theme_bw() +
  theme(axis.text.y = element_blank())
p
```

It doesn't make much difference.

It occurs to me that I've been specifying a gamma prior on $\theta$ with a shape and a rate (inverse scale) of 1. That leads to a prior mean of 1 and a prior variance of 1, which doesn't make a lot of sense, since in this model $\theta$ is $4N_e\mu$, where $\mu$ is the mutation rate per nucleotide site. Given that we expect $\mu \approx 10^{-8}$ a more reasonable (and still generous) prior mean for $\theta$ would be $10^{-2}$, corresponding to $N_e = 250,000$. If we leave the shape at 1 we need to make the scale $10^{-2}$ meaning that the rate should be 100. Let's give that a shot and see what we get. But first, I want to make sure I'm translating rate and scale correctly.

```{r}
scale_01 <- rgamma(10000, 1, scale = 0.01)
rate_100 <- rgamma(10000, 1, rate = 100)
for_plot <- tibble(x = c(scale_01, rate_100),
                   method = c(rep("Scale = 0.01", 10000), rep("Rate = 100", 10000)))
p <- ggplot(for_plot, aes(x = x, color = method, fill = method)) +
  geom_density(alpha = 0.2) +
  theme_bw()
p

for_plot_sum <- for_plot %>%
  group_by(method) %>%
  summarize(xbar = mean(x))
for_plot_sum
```
Looks good. The densities are indistinguishable, and the mean of each is very close to 0.01.


```{r}
results <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, n_rep = 100,
                           shape = 1, rate = 100, conditional = TRUE, ts_tv = 3)

p <- ggplot(results, aes(x = theta_B)) +
  geom_density() +
  geom_vline(xintercept = 0.0001, color = "salmon", linetype = "dashed") +
  theme_bw() +
  theme(axis.text.y = element_blank())
p
```

Even with 100 loci, the choice of prior makes a difference.

## Exactly one segregating site

In the simulations above I simply made sure there was a segregating site. For the ones that follow, I'm going to have exactly one segregating site.

```{r}
construct_a_finite <- function(dna_seqs, n_sample) {
  a <- numeric(n_sample)
  if (length(seg.sites(dna_seqs)) > 0 ) {
    n_seqs <- nrow(dna_seqs)
    dna_seqs_chr <- character(n_seqs)
    for (i in 1:n_seqs) {
      dna_seqs_chr[i] <- paste(dna_seqs[i, ], collapse = "")
    }
    ##
    ## need to figure out how to construct table of haplotype counts from table of 
    ## sequences
    ##
    counts <- table(as.factor(dna_seqs_chr))
    for (count in 1:n_sample) {
      index <- counts[count]
      a[index] <- a[index] + 1
    }
  } else {
    a[n_sample] <- 1
  }
  return(a)
}

## Run the simulation for a finite length sequence
##
## Note: ts_tv = 0.5 gives equal rates since there are 4 types of transversions and only 
## 2 types of transitions
##
simulate_finite <- function(n_sample, seq_len, n_loci, theta,  
                            n_rep = 1, shape = 1, rate = 1, conditional = FALSE,
                            ts_tv = 0.5) 
{
  theta_B <- numeric(n_rep)
  for (rep in 1:n_rep) {
    if (n_rep > 1) {
      cat(".")
      if (rep %% 5 == 0) {
        count <- sprintf("%4d", rep)
        cat(count)
      }
      if (rep %% 50 == 0) {
        cat("\n")
      }
    }
    ## simulate the data
    ##
    a <- matrix(nrow = n_loci, ncol = n_sample)
    for (l in 1:n_loci) {
      seg_sites <- 0
      ## if conditional == TRUE, keep constructing samples until we have a SNP
      ##
      while ((seg_sites != 1)) {
        ms_tree <- phyclust::ms(nsam = n_sample, nreps = 1, opts = paste("-T"))
        ## N.B.: seqs_chr comes back in PHYLIP format as a vector of strings (including the taxon)
        ## HKY with -t0.5 and without -f (frequencies) makes the model Jukes-Cantor
        ##
        options <- paste("-mHKY -l", seq_len, " -d", theta, " -t", ts_tv, " -q", sep = "")
        seqs_chr <- seqgen(opts = options, newick.tree = ms_tree[3])
        ## We write the result to disk
        ##
        write(seqs_chr, "seqs.phy")
        ## read it back using read.dna (from ape) and delete it
        ##
        tmp <- read.dna("seqs.phy")
        file.remove("seqs.phy")
        ## estimate theta using Watterson approach (theta_k)
        ##
        seg_sites <- length(seg.sites(tmp))
        if (conditional == FALSE) {
          break;
        }
      }
      ## construct a matrix for Bayesian estimate
      ##
      a[l, ] <- construct_a_finite(tmp, n_sample)
    }
    ## Bayesian estimate: single theta shared across all loci
    ##
    stan_data <- list(N = n_sample,
                      L = n_loci,
                      a = a,
                      shape = shape,
                      rate = rate) 
    if (conditional) {
      model <- "ewens-multilocus-conditional.stan"
    } else {
      model <- "ewens-multilocus.stan"
    }
    fit <- stan(file = model,
                data = stan_data,
                refresh = 0)
    fit_df <- as.data.frame(fit)
    theta_B[rep] <- mean(fit_df$theta, na.rm = TRUE)/seq_len
  }
  ret_val <- tibble(theta_B = theta_B)
  return(ret_val)
}
```


```{r warning = FALSE}
results_1 <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, n_rep = 100,
                             shape = 1, rate = 1, conditional = TRUE)
results_100 <- simulate_finite(n_sample = 25, seq_len = 250, n_loci = 100, theta = 0.0001, 
                               n_rep = 100,
                               shape = 1, rate = 100, conditional = TRUE)
for_plot <- tibble(rate = c(rep("Rate = 1", nrow(results_1)), rep("Rate = 100", nrow(results_1))),
                   theta = c(results_1$theta_B, results_100$theta_B))
p <- ggplot(for_plot, aes(x = theta, color = rate, fill = rate)) +
  geom_density(alpha = 0.2) +
  geom_vline(xintercept = 0.0001, color = "salmon", linetype = "dashed") +
  theme_bw() +
  theme(axis.text.y = element_blank())
p
```
