---
title: "Multilocus estimates of pi"
output: html_notebook
---

# Overview

We'll simulate $L$ independently inherited loci, each B base pairs in length. We'll estimate $\theta$ from the number of segregating sites, $\theta_K$, from the nucleotide diversity, $\theta_pi$, and from Ewen's sampling formula, $\theta_B$, where we assume the same $\theta_B$ across all loci.

## Load libraries and clear out memory

```{r}
library(phyclust)
library(rstan)
library(tidyverse)
library(ggplot2)

rm(list = ls())

options(mc.cores = parallel::detectCores())
rstan_options(auto_write = TRUE)
```

## Utility functions

```{r}
## calculate a harmonic sum: \sum_i 1/i
##
harm_sum <- function(n) {
  k <- 0
  for (i in 1:(n-1)) {
    k <- k + 1/i
  }
  return(k)
}

## Watterson's estimate of theta
##
est_theta_k <- function(seg_sites, n_sample, seq_len) {
  return(length(seg_sites)/(harm_sum(n_sample)*seq_len))
}

## nucleotide diversity estimate of theta
##
est_theta_pi <- function(dna_seqs) {
  n_seqs <- dim(dna_seqs)[1]
  seq_length <- dim(dna_seqs)[2]
  pi <- 0.0 
  for (i in 1:(n_seqs - 1)) {
    for (j in (i+1):n_seqs) {
      pi <- pi + length(seg.sites(dna_seqs[c(i, j), ]))/seq_length
    }
  }
  pi <- pi/(n_seqs*(n_seqs - 1)/2)
  return(pi)
}

## construct matrix of allele counts, loci in rows
##
construct_a <- function(dna_seqs, n_sample) {
  a <- numeric(n_sample)
  if (length(seg.sites(dna_seqs)) > 0 ) {
    n_seqs <- nrow(dna_seqs)
    dna_seqs_chr <- character(n_seqs)
    for (i in 1:n_seqs) {
      dna_seqs_chr[i] <- paste(dna_seqs[i, ], collapse = "")
    }
    ##
    ## need to figure out how to construct table of haplotype counts from table of 
    ## sequences
    ##
    counts <- table(as.factor(dna_seqs_chr))
    for (count in 1:n_sample) {
      index <- counts[count]
      a[index] <- a[index] + 1
    }
  } else {
    a[ncol(dna_seqs)] <- 1
  }
  return(a)
}

## Run the simulation
##
simulate <- function(n_sample, seq_len, n_loci, theta, n_rep = 1,
                     shape = 1, rate = 1) 
{
  theta_k<- numeric(n_rep)
  theta_pi <- numeric(n_rep)
  theta_B <- numeric(n_rep)
  for (rep in 1:n_rep) {
    if (n_rep > 1) {
      cat(".")
      if (rep %% 5 == 0) {
        count <- sprintf("%4d", rep)
        cat(count)
      }
      if (rep %% 50 == 0) {
        cat("\n")
      }
    }
    ## simulate the data
    ##
    theta_kl <- numeric(n_loci)
    theta_pil <- numeric(n_loci)
    a <- matrix(nrow = n_loci, ncol = n_sample)
    for (l in 1:n_loci) {
      ms_tree <- phyclust::ms(nsam = n_sample, nreps = 1, opts = paste("-T"))
      ## N.B.: seqs_chr comes back in PHYLIP format as a vector of strings (including the taxon)
      ## HKY without -t (transitiion/transversion ratio) and without -f (frequencies) makes the
      ## model Jukes-Cantor
      ##
      options <- paste("-mHKY -l", seq_len, " -d", theta, " -q", sep = "")
      seqs_chr <- seqgen(opts = options, newick.tree = ms_tree[3])
      ## We write the result to disk
      ##
      write(seqs_chr, "seqs.phy")
      ## read it back using read.dna (from ape) and delete it
      ##
      tmp <- read.dna("seqs.phy")
      file.remove("seqs.phy")
      ## estimate theta using Watterson approach (theta_k)
      ##
      seg_sites <- seg.sites(tmp)
      theta_kl[l] <- est_theta_k(seg_sites, n_sample, seq_len)
      ## estimate theta using nucleotide diversity (theta_pi)
      ##
      theta_pil[l] <- est_theta_pi(tmp)
      ## construct a matrix for Bayesian estimate
      a[l, ] <- construct_a(tmp, n_sample)
    }
    theta_k[rep] <- mean(theta_kl)
    theta_pi[rep] <- mean(theta_pil)
    ## Bayesian estimate: single theta shared across all loci
    ##
    stan_data <- list(N = n_sample,
                      L = n_loci,
                      a = a,
                      shape = shape,
                      rate = rate) 
    fit <- stan(file = "ewens-multilocus.stan",
                data = stan_data,
                refresh = 0)
    fit_df <- as.data.frame(fit)
    theta_B[rep] <- mean(fit_df$theta, na.rm = TRUE)/seq_len
  }
  ret_val <- tibble(theta_k = theta_k,
                    theta_pi = theta_pi,
                    theta_B = theta_B)
  return(ret_val)
}
```

## Run a simulation

```{r}
results <- simulate(n_sample = 25,
                    seq_len = 1000,
                    n_loci = 100,
                    theta = 0.1,
                    n_rep = 25)

for_plot <- pivot_longer(results, cols = everything(), names_to = "Model")
p <- ggplot(for_plot, aes(x = value, fill = Model, color = Model)) +
  geom_density(alpha = 0.2) +
  theme_bw()
p
```

```{r}
results <- simulate(n_sample = 25,
                    seq_len = 1000,
                    n_loci = 100,
                    theta = 1.0,
                    n_rep = 25)

for_plot <- pivot_longer(results, cols = everything(), names_to = "Model")
p <- ggplot(for_plot, aes(x = value, fill = Model, color = Model)) +
  geom_density(alpha = 0.2) +
  theme_bw()
p
```


```{r}
results <- simulate(n_sample = 250,
                    seq_len = 1000,
                    n_loci = 100,
                    theta = 1.0,
                    n_rep = 25)

for_plot <- pivot_longer(results, cols = everything(), names_to = "Model")
p <- ggplot(for_plot, aes(x = value, fill = Model, color = Model)) +
  geom_density(alpha = 0.2) +
  theme_bw()
p
```

